### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by MancunianSam**

### CODE REVIEW

File: Jenkinsfile-deploy [hunk 1]
```diff
@@ -0,0 +1,35 @@
+pipeline {
+    agent {
+        label "master"
+    }
+    parameters {
+        choice(name: "STAGE", choices: ["intg", "staging", "prod"], description: "The stage you are building the front end for")
```
Comment (TomJKing) [inline Jenkinsfile-deploy:6]: Probably want to change the description in the parameters section

File: Jenkinsfile-deploy [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        stage("Deploy lambda") {
+            agent {
+                ecs {
+                    inheritFrom "aws"
+                    taskrole "arn:aws:iam::${env.MANAGEMENT_ACCOUNT}:role/TDRJenkinsNodeLambdaRole${params.STAGE.capitalize()}"
+                }
+            }
+            steps {
+                script {
+                    def accountNumber = getAccountNumberFromStage()
```
Comment (TomJKing) [inline Jenkinsfile-deploy:19]: `getAccountNumberFromStage` function is available in the shared library

File: Jenkinsfile-test [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+def versionTag = "v${env.BUILD_NUMBER}"
+def repo = "tdr-api-update"
+
+pipeline {
+    agent {
+        label "master"
+    }
+    parameters {
+        choice(name: "STAGE", choices: ["intg", "staging", "prod"], description: "The stage you are building the front end for")
```
Comment (TomJKing) [inline Jenkinsfile-test:11]: Probably want to change the description in the parameters section

File: build.sbt [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import Dependencies._
+import sbt.Keys.fork
+
+ThisBuild / scalaVersion := "2.13.2"
+ThisBuild / version := "0.1.0-SNAPSHOT"
```
Comment (TomJKing) [inline build.sbt:5]: Is the SNAPSHOT build correct?

File: build.sbt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+fork in Test := true
+javaOptions in Test += s"-Dconfig.file=${sourceDirectory.value}/test/resources/application.conf"
+
+assemblyMergeStrategy in assembly := {
+  case PathList("META-INF", xs@_*) => MergeStrategy.discard
+  case _ => MergeStrategy.first
+}
+resolvers += "TDR Releases" at "s3://tdr-releases-mgmt"
+assemblyJarName in assembly := "api-update.jar"
```
Comment (TomJKing) [inline build.sbt:35]: Missing line break at end of file

File: src/main/resources/application.conf [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    id = "tdr-backend-checks"
+    id = ${?CLIENT_ID}
+    secret = "secretplaceholder"
+    secret = ${?CLIENT_SECRET}
+}
+sqs {
+    endpoint = "https://sqs.eu-west-2.amazonaws.com"
+    url = "set_me_queue_url"
+    url = ${?QUEUE_URL}
+}
```
Comment (TomJKing) [inline src/main/resources/application.conf:17]: Missing line break at end of file

File: src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package uk.gov.nationalarchives.api.update
+
+import com.amazonaws.services.lambda.runtime.Context
+import com.amazonaws.services.lambda.runtime.events.SQSEvent
+import graphql.codegen.AddAntivirusMetadata.{AddAntivirusMetadata => av}
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala:5]: Maybe change import abbreviation to: avm

File: src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            processor.process(avInput, bodyWithReceiptHandle.recieptHandle)
+          case fileMetadataInput: AddFileMetadataInput =>
+            println(s"file property name ${fileMetadataInput.filePropertyName}")
+            val processor = new Processor[AddFileMetadataInput, afm.Data, afm.Variables](afm.document, i => afm.Variables(i))
+            processor.process(fileMetadataInput, bodyWithReceiptHandle.recieptHandle)
+        }
+      }).toList
+
+    Await.result(ResultCollector().collect(results), 10 seconds)
+
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala:37]: Unnecessary line break

File: src/main/scala/uk/gov/nationalarchives/api/update/ResultCollector.scala [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+      } else {
+        successes
+      }
+    })
+  }
+}
+
+object ResultCollector {
+  def apply()(implicit executionContext: ExecutionContext): ResultCollector = new ResultCollector()
+}
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/api/update/ResultCollector.scala:25]: Missing line break at end of file

File: src/test/resources/application.conf [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+}
+client {
+    id = "id"
+    secret = "secret"
+}
+
+sqs {
+    url = "http://localhost:8001/1/testqueue"
+    endpoint = "http://localhost:8001"
+}
```
Comment (TomJKing) [inline src/test/resources/application.conf:13]: Missing line break at end of file

File: src/test/scala/uk/gov/nationalarchives/api/update/ApiUpdateTest.scala [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    val client = mock[GraphQLClient[Data, Variables]]
+    val document = mock[Document]
+    val keycloakUtils = mock[KeycloakUtils]
+
+    when(keycloakUtils.serviceAccountToken(any[String], any[String]))
+      .thenReturn(Future.successful(new BearerAccessToken("token")))
+    when(client.getResult(any[BearerAccessToken], any[Document], any[Option[Variables]]))
+      .thenReturn(Future.successful(GraphQlResponse(Some(Data(TestResponse())), List())))
+    apiUpdate.send(keycloakUtils, client, document, Variables()).futureValue
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/api/update/ApiUpdateTest.scala:35]: Unnecessary line break

File: src/test/scala/uk/gov/nationalarchives/api/update/ApiUpdateTest.scala [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    val variables = Variables()
+
+    when(keycloakUtils.serviceAccountToken(any[String], any[String]))
+      .thenReturn(Future.successful(new BearerAccessToken("token")))
+    when(client.getResult(any[BearerAccessToken], any[Document], any[Option[Variables]]))
+      .thenReturn(Future.successful(GraphQlResponse(Some(Data(TestResponse())), List())))
+    apiUpdate.send(keycloakUtils, client, document, variables).futureValue
+
+    verify(client).getResult(new BearerAccessToken("token"), document, Some(variables))
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/api/update/ApiUpdateTest.scala:60]: Unnecessary line break

File: src/test/scala/uk/gov/nationalarchives/api/update/ChecksumUpdateTest.scala [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package uk.gov.nationalarchives.api.update
+
+import uk.gov.nationalarchives.api.update.utils.ExternalServicesTest
+
+class ChecksumUpdateTest extends ExternalServicesTest {
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/api/update/ChecksumUpdateTest.scala:6]: Is this a placeholder file, as there are no tests included?

File: src/test/scala/uk/gov/nationalarchives/api/update/ResultCollectorTest.scala [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    caught.getMessage should endWith("decoderError.")
+  }
+
+  "The collect method" should "return successfully if all responses have succeeded" in {
+    val responseProcessor = ResultCollector()
+    val id = "successful-id"
+    val processed = responseProcessor.collect(List(Right(Future.successful(Right(id)))))
+    processed.futureValue.head should equal(id)
+  }
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/api/update/ResultCollectorTest.scala:40]: Unnecessary line break

File: src/test/scala/uk/gov/nationalarchives/api/update/UpdateTest.scala [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+      authOkJson("access_token")
+      graphqlOkJson(s"graphql_valid_${updateType}_response")
+      val main = new Lambda()
+      Try(main.update(event, context))
+      val messages = client.receiveMessage(ReceiveMessageRequest.builder.queueUrl(queueUrl).build)
+      messages.hasMessages should be(true)
+      messages.messages.size should be(1)
+    }
+  }
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/api/update/UpdateTest.scala:78]: Indentation of braces

File: src/test/scala/uk/gov/nationalarchives/api/update/utils/ExternalServicesTest.scala [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+      ReceiveMessageRequest
+        .builder
+        .queueUrl(queueUrl)
+        .maxNumberOfMessages(10)
+        .build
+    ).messages.asScala.foreach(msg => {
+      client.deleteMessage(DeleteMessageRequest.builder.queueUrl(queueUrl).receiptHandle(msg.receiptHandle).build())
+    })
+  }
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/api/update/utils/ExternalServicesTest.scala:84]: Unnecessary line breaks

File: Jenkinsfile-deploy (hunk 1 above)
Comment (MancunianSam) [reply Jenkinsfile-deploy:6]: Yeah good point. I've changed the other file as well.

File: Jenkinsfile-deploy (hunk 2 above)
Comment (MancunianSam) [reply Jenkinsfile-deploy:19]: I've changed it.

File: build.sbt (hunk 4 above)
Comment (MancunianSam) [reply build.sbt:5]: No, probably not. If this was a library we were releasing then yeah but it's not so I've changed it to 0.1.0

File: src/test/scala/uk/gov/nationalarchives/api/update/utils/ExternalServicesTest.scala (hunk 16 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/api/update/utils/ExternalServicesTest.scala:84]: I've fixed each of the line break comments.

File: src/test/scala/uk/gov/nationalarchives/api/update/UpdateTest.scala (hunk 15 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/api/update/UpdateTest.scala:78]: I think this is better now, have a look.

File: src/main/resources/application.conf (hunk 6 above)
Comment (MancunianSam) [reply src/main/resources/application.conf:17]: I've added this in.

File: src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala (hunk 7 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala:5]: Makes sense, I've changed it.

File: src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala (hunk 8 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/api/update/Lambda.scala:37]: I've changed all of the line break ones in this bit as well.

File: Jenkinsfile-deploy [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+        stage("Deploy lambda") {
+            agent {
+                ecs {
+                    inheritFrom "aws"
+                    taskrole "arn:aws:iam::${env.MANAGEMENT_ACCOUNT}:role/TDRJenkinsNodeLambdaRole${params.STAGE.capitalize()}"
+                }
+            }
+            steps {
+                script {
+                    def accountNumber = tdr.getAccountNumberFromStage()
```
Comment (TomJKing) [inline Jenkinsfile-deploy:21]: Missing the stage param for the `tdr.getAccountNumberFromStage()` function

File: Jenkinsfile-deploy (hunk 2 above)
Comment (TomJKing) [reply Jenkinsfile-deploy]: Looks like the stage param for the `tdr.getAccountNumberFromStage()` function that you added

File: Jenkinsfile-deploy (hunk 17 above)
Comment (MancunianSam) [reply Jenkinsfile-deploy]: One of these days I will remember this :smile:

**Pull request merged**